### PR TITLE
Overhaul Pages and Bytes types

### PR DESCRIPTION
**Pull request by Robbepop**
Problems with previous types were found during fuzzy testing.
These types make it simpler to fix linear memory allocation problems in `wasmi` for 32-bit and 64-bit architectures for edge cases.

Also we get rid of the `memory_units` dependency which did a lot more things that we actually needed in `wasmi`.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## CRITERION BENCHMARKS ## 

 |BENCHMARK|MASTER|PR|Diff|
 |---|---|---|---|
 | `compile_and_validate_v1`  | 5.5011 ms  | 5.4499 ms  | :white_circle:  -0.8252% |
| `execute_count_until_v1`  | 2.0224 ms  | 1.9880 ms  | :white_circle:  -1.3218% |
| `execute_factorial_iterative_v1`  | 920.26 ns  | 965.89 ns  | :red_circle:  +5.0415% |
| `execute_factorial_recursive_v1`  |  1.2443 µs   |  1.2750 µs   | :red_circle:  +2.4554% |
| `execute_fib_iterative_v1`  | 4.8087 ms  | 4.8859 ms  | :red_circle:  +1.6495% |
| `execute_fib_recursive_v1`  | 11.140 ms  | 11.521 ms  | :red_circle:  +3.4126% |
| `execute_global_bump_v1`  | 2.8159 ms  | 2.7106 ms  | :green_circle:  -4.0504% |
| `execute_host_calls_v1`  |  30.944 µs   |  31.985 µs   | :red_circle:  +3.4415% |
| `execute_memory_fill_v1`  | 4.0582 ms  | 4.0828 ms  | :white_circle:  +0.6461% |
| `execute_memory_sum_v1`  | 3.8010 ms  | 3.8462 ms  | :red_circle:  +1.1487% |
| `execute_memory_vec_add_v1`  | 8.0508 ms  | 8.0662 ms  | :white_circle:  +0.1672% |
| `execute_recursive_is_even_v1`  | 2.2377 ms  | 2.2760 ms  | :red_circle:  +1.6090% |
| `execute_recursive_ok_v1`  |  295.42 µs   |  300.36 µs   | :red_circle:  +1.6719% |
| `execute_recursive_scan_v1`  |  356.99 µs   |  366.62 µs   | :red_circle:  +2.7166% |
| `execute_recursive_trap_v1`  |  25.476 µs   |  25.968 µs   | :red_circle:  +2.2106% |
| `execute_regex_redux_v1`  | 1.3963 ms  | 1.4004 ms  | :white_circle:  +0.4005% |
| `execute_rev_complement_v1`  | 1.3919 ms  | 1.3978 ms  | :white_circle:  +0.7223% |
| `execute_tiny_keccak_v1`  | 1.2336 ms  | 1.2493 ms  | :red_circle:  +1.3615% |
| `execute_trunc_f2i_v1`  | 1.8026 ms  | 1.8128 ms  | :white_circle:  +0.5401% |
| `instantiate_v1`  |  73.530 µs   |  73.327 µs   | :white_circle:  -0.3675% |


 [Link to pipeline](https://gitlab.parity.io/parity/mirrors/wasmi/-/jobs/1853108)

File: wasmi_v1/src/memory/byte_buffer.rs [hunk 1]
```diff
@@ -14,30 +15,59 @@ pub struct ByteBuffer {
...
+    }
+
+    #[cfg(target_pointer_width = "32")]
+    const fn max_len() -> u64 {
+        i32::MAX as u32 as u64 + 1
+    }
+
+    #[cfg(target_pointer_width = "64")]
+    const fn max_len() -> u64 {
+        u32::MAX as u64 + 1
```
Comment (yjhmelody) [inline wasmi_v1/src/memory/byte_buffer.rs:30]: Sorry, I could find this value in spec. Could you give me a ref?

**Comment by yjhmelody** [comment]
Could you also consider to extract memoryEntity type as a new crate?

**Comment by Robbepop** [comment]
> Could you also consider to extract memoryEntity type as a new crate?

i do not see a reason for this. also what does this has to do with this PR?

**Comment by yjhmelody** [comment]
> > Could you also consider to extract memoryEntity type as a new crate?
> 
> i do not see a reason for this. also what does this has to do with this PR?

I think it may be better to put Pages/Bytes into a `wasmi_memory`  crate.

**Comment by Robbepop** [comment]
> > > Could you also consider to extract memoryEntity type as a new crate?
> > 
> > 
> > i do not see a reason for this. also what does this has to do with this PR?
> 
> I think it may be better to put Pages/Bytes into a `wasmi_memory` crate.

Yeah I got that but can you give me a compelling example as to why?

**Comment by Robbepop** [comment]
> > Yeah I got that but can you give me a compelling example as to why?
> 
> No examples, simply because I need this crate too :(

Cannot you use `wasmi_core` instead?

**Pull request merged**